src/CsvHelper/Configuration/ConfigurationFunctions.cs

```diff
@@ -197,24 +199,38 @@ public static string GetDynamicPropertyName(GetDynamicPropertyNameArgs args)
 			}
 
 			var newLine = config.NewLine;
-			if ((new[] { "\r\n", "\r", "\n" }).Contains(newLine))
+			if (lineEndings.Contains(newLine))
 			{
 				newLine = "\r\n|\r|\n";
 			}
 
 			var lineDelimiterCounts = new List<Dictionary<string, int>>();
+			var delimitersToCheck = new List<string>(config.DetectDelimiterValues);
+
 			while (text.Length > 0)
 			{
 				// Since all escaped text has been removed, we can reliably read line by line.
 				var match = Regex.Match(text, newLine);
 				var line = match.Success ? text.Substring(0, match.Index + match.Length) : text;
 
 				var delimiterCounts = new Dictionary<string, int>();
-				foreach (var delimiter in config.DetectDelimiterValues)
+				var i = 0;
+				while (i < delimitersToCheck.Count)
 				{
+					var delimiter = delimitersToCheck[i];
 					// Escape regex special chars to use as regex pattern.
-					var pattern = Regex.Replace(delimiter, @"([.$^{\[(|)*+?\\])", "\\$1");
-					delimiterCounts[delimiter] = Regex.Matches(line, pattern).Count;
+					var pattern = specialCharsRegex.Replace(delimiter, "\\$1");
```

Comment:
```suggestion
					var pattern = Regex.Escape(delimiter);
```
? Though there is not a whole lot of test coverage for delimiters containing special regex characters

Reply:
Fixed after verifying no behavior change (in PowerShell).

```powershell
> '.$^{\[(|)*+?\'.ToCharArray() | % { [regex]::Escape($_)}

\.
\$
\^
\{
\\
\[
\(
\|
\)
\*
\+
\?
\\
```

Reply:
I also added a [test for `+` as delimiter](https://github.com/JoshClose/CsvHelper/pull/2129/files#diff-73a3d3bfe9b3603478f83fabdddae43f6ebbdacf4d8d5d6f7c15fff5e58ff81fR69-R86).